### PR TITLE
Lamb optimizer update

### PR DESCRIPTION
**Pull request by access2rohit**
## Description ##
adding to new operators:
- lamb_update_phase1
- lamb_update_phase2
Link to paper: https://arxiv.org/pdf/1904.00962.pdf

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] lamb_update, tests, (and when applicable, API doc)

### Testing ###
```
[DEBUG] 1000 of 1000: Setting test np/mx/python random seeds, use MXNET_TEST_SEED=2052238159 to reproduce.
ok

----------------------------------------------------------------------
Ran 1 test in 5085.147s

OK
```

### CODE REVIEW

**Comment by access2rohit** [comment]
@mxnet-label-bot add [pr-ready-to-review]

File: python/mxnet/optimizer/optimizer.py [hunk 1]
```diff
@@ -1244,6 +1244,51 @@ def update(self, index, weight, grad, state):
             kwargs = {}
             sgd_update(weight, grad, out=weight, lr=lr, wd=wd, **kwargs)
 
+
+@register
+class LAMB(Optimizer):
+    """LAMB Optimizer.
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer/optimizer.py]: pls add doc

File: python/mxnet/optimizer/optimizer.py [hunk 2]
```diff
@@ -1244,6 +1244,51 @@ def update(self, index, weight, grad, state):
...
+                  'bias_correction': self.bias_correction, 't': t,
+                  'rescale_grad': self.rescale_grad}
+        if self.clip_gradient:
+            kwargs['clip_gradient'] = self.clip_gradient
+
+        mean, var = state
+        g = lamb_update(weight, grad, mean, var, lr=lr, wd=wd, **kwargs)
+        r_1 = weight.norm()
+        r_2 = g.norm()
+        lamb_weight_update(weight, g, lr, r_1.asnumpy()[0], r_2.asnumpy()[0], self.lower_bound,
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer/optimizer.py]: we cannot have asnumpy in the implementation. it's slow

File: src/operator/optimizer_op.cc [hunk 3]
```diff
@@ -921,5 +923,33 @@ Note that non-zero values for the weight decay option are not supported.
...
+.set_attr<FCompute>("FCompute<cpu>", LAMBUpdate<cpu>)
+.add_argument("weight", "NDArray-or-Symbol", "Weight")
+.add_argument("grad", "NDArray-or-Symbol", "Gradient")
+.add_argument("mean", "NDArray-or-Symbol", "Moving mean")
+.add_argument("var", "NDArray-or-Symbol", "Moving variance")
+.add_arguments(LAMBParam::__FIELDS__());
+
+NNVM_REGISTER_OP(lamb_weight_update)
+.describe(R"code(Update function for lamb optimizer.
+)code" ADD_FILELINE)
```
Comment (eric-haibin-lin) [inline src/operator/optimizer_op.cc]: please specify FMutateInput..

File: src/operator/optimizer_op-inl.h [hunk 4]
```diff
@@ -1563,6 +1563,186 @@ inline void AdamUpdateEx(const nnvm::NodeAttrs& attrs,
...
+
+struct LAMBUpdateKernel {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* out_data,
+    DType* mean_data, DType* var_data, const DType* weight_data, const DType* grad_data,
+    const DType clip_gradient, const DType rescale_grad,
+    const DType beta1, const DType beta2,
+    DType lr, const DType wd,
+    const DType epsilon, const DType lower_bound,
+    const DType upper_bound, const DType t,
```
Comment (eric-haibin-lin) [inline src/operator/optimizer_op-inl.h]: are upper_bound, lower_bound, etc being used at all??? Remove them from the argument list if they are not required

File: tests/python/unittest/test_optimizer.py [hunk 5]
```diff
@@ -425,6 +425,77 @@ def test_nag():
...
+
+@with_seed()
+def test_lamb():
+    opt1 = PyLAMB
+    opt2 = mx.optimizer.LAMB
+    shape = (3, 4, 5)
+    cg_options = [{}, {'clip_gradient': 0.4}, {'clip_gradient': 0.5}]
+    rg_options = [{}, {'rescale_grad': 0.14}, {'rescale_grad': 0.8}]
+    wd_options = [{}, {'wd': 0.03}, {'wd': 0.05}, {'wd': 0.07}]
+    bc_options = [{}, {'bias_correction': False}, {'bias_correction': True}]
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_optimizer.py]: no test for upper / lower bound?

File: tests/python/unittest/test_optimizer.py [hunk 6]
```diff
@@ -425,6 +425,77 @@ def test_nag():
...
+        r1 = weight.norm()
+        if not self.bias_correction:
+            r1 = mx.nd.minimum(mx.nd.maximum(r1, self.lower_bound), self.upper_bound)
+            g = mean / (mx.nd.sqrt(var) + self.epsilon) + wd * weight
+
+        else:
+            mean_hat = mean / (1. - mx.nd.power(self.beta1, t))
+            var_hat = var / (1. - mx.nd.power(self.beta2, t))
+            g = mean_hat / mx.nd.sqrt(var_hat + self.epsilon) + wd * weight
+
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_optimizer.py]: can we update the python implementation to be:
```
            if self.lower_bound:
                r1 = maximum(r1, self.lower_bound)
            if self.upper_bound:
                r1 = minimum(r1, self.upper_bound)
            if self.bias_correction:
                mean_hat = mean / (1. - power(self.beta1, t))
                var_hat = var / (1. - power(self.beta2, t))
            else:
                mean_hat = mean
                var_hat = var
            g = mean_hat / (sqrt(var_hat) + self.epsilon) + wd * weight

```
and update the c++ version accordingly?

File: src/operator/optimizer_op.cc (hunk 3 above)
Comment (access2rohit) [reply src/operator/optimizer_op.cc]: Done

File: tests/python/unittest/test_optimizer.py (hunk 5 above)
Comment (access2rohit) [reply tests/python/unittest/test_optimizer.py]: done

File: tests/python/unittest/test_optimizer.py (hunk 6 above)
Comment (access2rohit) [reply tests/python/unittest/test_optimizer.py]: done

File: python/mxnet/optimizer/optimizer.py (hunk 1 above)
Comment (access2rohit) [reply python/mxnet/optimizer/optimizer.py]: working on it now

File: src/operator/optimizer_op.cc (hunk 3 above)
Comment (access2rohit) [reply src/operator/optimizer_op.cc]: done here: https://github.com/apache/incubator-mxnet/pull/16715/files#diff-abf63a5538d1f9e37ec5d95d7528d2d0R935

File: src/operator/optimizer_op.cc [hunk 7]
```diff
@@ -921,5 +923,39 @@ Note that non-zero values for the weight decay option are not supported.
...
+.set_num_inputs(4)
+.set_num_outputs(1)
+.set_attr_parser(ParamParser<LAMBWeightParam>)
+.set_attr<mxnet::FInferShape>("FInferShape", LambWeightShape)
+.set_attr<nnvm::FInferType>("FInferType", ElemwiseType<4, 1>)
+.set_attr<FCompute>("FCompute<cpu>", LAMBWeightUpdate<cpu>)
+.add_argument("weight", "NDArray-or-Symbol", "Weight")
+.add_argument("g", "NDArray-or-Symbol", "g")
```
Comment (eric-haibin-lin) [inline src/operator/optimizer_op.cc]: need better documentation for `g`

File: tests/python/unittest/test_optimizer.py [hunk 8]
```diff
@@ -425,6 +425,79 @@ def test_nag():
...
             compare_optimizer(opt1(**kwarg), opt2(**kwarg), shape, dtype, rtol=1e-3, atol=1e-4)
 
+
+# LAMB optimizer
+class PyLAMB(mx.optimizer.Optimizer):
+    """
+	Python reference implementation of LAMB optimizer.
+    """
+    def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-6,
+                 lower_bound=1e-3, upper_bound=10.0, bias_correction=False, **kwargs):
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_optimizer.py]: inconsistent default values for `lower_bound` and `upper_bound` compared to mx.optim.LAMB

File: tests/python/unittest/test_optimizer.py (hunk 5 above)
Comment (eric-haibin-lin) [reply tests/python/unittest/test_optimizer.py]: we should also test lower_bound:1e-3 and upper_bound: 10

File: src/operator/optimizer_op-inl.h [hunk 9]
```diff
@@ -1563,6 +1563,192 @@ inline void AdamUpdateEx(const nnvm::NodeAttrs& attrs,
...
+};
+
+struct LAMBWeightParam : public dmlc::Parameter<LAMBWeightParam> {
+    float lr;
+    float lower_bound;
+    float upper_bound;
+    DMLC_DECLARE_PARAMETER(LAMBWeightParam) {
+      DMLC_DECLARE_FIELD(lr)
+      .describe("Learning rate");
+      DMLC_DECLARE_FIELD(lower_bound)
```
Comment (eric-haibin-lin) [inline src/operator/optimizer_op-inl.h]: The default value is inconsistent with how we deal with wd

Comment (access2rohit) [reply src/operator/optimizer_op-inl.h]: done

File: tests/python/unittest/test_optimizer.py (hunk 5 above)
Comment (access2rohit) [reply tests/python/unittest/test_optimizer.py]: done

File: src/operator/optimizer_op.cc (hunk 7 above)
Comment (access2rohit) [reply src/operator/optimizer_op.cc]: Can you suggest a link or what should I write here. I have followed this: https://towardsdatascience.com/an-intuitive-understanding-of-the-lamb-optimizer-46f8c0ae4866
Not sure how should I describe `g`

File: src/operator/optimizer_op-inl.h [hunk 10]
```diff
@@ -1563,6 +1563,192 @@ inline void AdamUpdateEx(const nnvm::NodeAttrs& attrs,
...
+                       const std::vector<TBlob> &outputs) {
+  using namespace mxnet_op;
+  const LambUpdatePhaseOneParam& param = nnvm::get<LambUpdatePhaseOneParam>(attrs.parsed);
+  Stream<xpu>* s = ctx.get_stream<xpu>();
+  MSHADOW_REAL_TYPE_SWITCH(inputs[0].type_flag_, DType, {
+    Tensor<xpu, 2, DType> weight = inputs[0].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> grad = inputs[1].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> mean = inputs[2].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> var = inputs[3].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> out = outputs[0].FlatTo2D<xpu, DType>(s);
```
Comment (samskalicky) [inline src/operator/optimizer_op-inl.h]: Are these assignments data copies? If so shouldnt we be using ResourceManager to allocate temp space?

File: src/operator/optimizer_op-inl.h [hunk 11]
```diff
@@ -1563,6 +1563,192 @@ inline void AdamUpdateEx(const nnvm::NodeAttrs& attrs,
...
+                       const std::vector<TBlob> &outputs) {
+  using namespace mxnet_op;
+  const LambUpdatePhaseTwoParam& param = nnvm::get<LambUpdatePhaseTwoParam>(attrs.parsed);
+  Stream<xpu>* s = ctx.get_stream<xpu>();
+  MSHADOW_REAL_TYPE_SWITCH(inputs[0].type_flag_, DType, {
+    Tensor<xpu, 2, DType> weight = inputs[0].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> g = inputs[1].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> r1 = inputs[2].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> r2 = inputs[3].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> out = outputs[0].FlatTo2D<xpu, DType>(s);
```
Comment (samskalicky) [inline src/operator/optimizer_op-inl.h]: Same here, are these copies?

**Comment by larroy** [comment]
Please add description and reference to paper in the PR.

File: python/mxnet/optimizer/optimizer.py (hunk 1 above)
Comment (larroy) [reply python/mxnet/optimizer/optimizer.py]: The name is clashing with the GLuon one, can we give it a different name?

**Comment by larroy** [comment]
I see a crash in the way gluon trainer is calling the optimizer...
```
+ python3 run_pretraining.py '--data=/home/piotr/mxnet-data/bert-pretraining/datasets/book-corpus/book-corpus-large-split/*.train,/home/piotr/mxnet-data/bert-pretraining/datasets/enwiki/enwiki-feb-doc-split/*.train' '--data_eval=/home/piotr/mxnet-data/bert-pretraining/datasets/book-corpus/book-corpus-large-split/*.test,/home/piotr/mxnet-data/bert-pretraining/datasets/enwiki/enwiki-feb-doc-split/*.test' --optimizer lamb3 --warmup_ratio 0.2 --num_steps 200 --ckpt_interval 300000000 --dtype float16 --ckpt_dir ./test-ckpt --lr 0.0001 --total_batch_size 32 --total_batch_size_eval 32 --accumulate 1 --model bert_24_1024_16 --max_seq_length 128 --max_predictions_per_seq 20 --num_data_workers 1 --eval_interval 100000000 --verbose --no_compute_acc --raw --comm_backend horovod --log_interval 10 --verbose --synthetic_data --raw --eval_use_npz
[22:42:38] ../src/storage/storage.cc:110: Using GPUPooledRoundedStorageManager.
Traceback (most recent call last):
  File "run_pretraining.py", line 574, in <module>
    train(data_train, data_eval, model)
  File "run_pretraining.py", line 457, in train
    num_ctxs=len(ctxs) * num_workers)
  File "/home/piotr/gluon-nlp/scripts/bert/fp16_utils.py", line 433, in step
    self.fp32_trainer.update(step_size)
  File "/home/piotr/mxnet_lamb/python/mxnet/gluon/trainer.py", line 397, in update
    self._update(ignore_stale_grad)
  File "/home/piotr/mxnet_lamb/python/mxnet/gluon/trainer.py", line 434, in _update
    updater(i, w, g)
  File "/home/piotr/mxnet_lamb/python/mxnet/optimizer/optimizer.py", line 1777, in __call__
    self.optimizer.update_multi_precision(i, w, g, self.states[i])
  File "/home/piotr/mxnet_lamb/python/mxnet/optimizer/optimizer.py", line 291, in update_multi_precision
    self.update(index, weight_master_copy, grad32, original_state)
  File "/home/piotr/mxnet_lamb/python/mxnet/optimizer/optimizer.py", line 1012, in update
    g = lamb_update(weight, grad, mean, var, wd=wd, **kwargs)
  File "<string>", line 88, in lamb_update
  File "/home/piotr/mxnet_lamb/python/mxnet/_ctypes/ndarray.py", line 107, in _imperative_invoke
    ctypes.byref(out_stypes)))
  File "/home/piotr/mxnet_lamb/python/mxnet/base.py", line 254, in check_call
    raise MXNetError(py_str(_LIB.MXGetLastError()))
mxnet.base.MXNetError: Some trailing characters could not be parsed: '[0.0015625]
<NDArray 1 @gpu(0)>', in operator lamb_update(name="", rescale_grad="
[0.0015625]
<NDArray 1 @gpu(0)>", wd="0.01", bias_correction="True", t="1", epsilon="1e-06", beta2="0.999", beta1="0.9")
```

**Comment by access2rohit** [comment]
> Please add description and reference to paper in the PR.

Done

**Comment by larroy** [comment]
@access2rohit could you answer Sam's comments?

File: python/mxnet/optimizer/optimizer.py [hunk 12]
```diff
@@ -1244,6 +1244,54 @@ def update(self, index, weight, grad, state):
             kwargs = {}
             sgd_update(weight, grad, out=weight, lr=lr, wd=wd, **kwargs)
 
+
+@register
+class LAMB(Optimizer):
+    """LAMB Optimizer.
+    """
+    def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-6,
+                 lower_bound=None, upper_bound=None, bias_correction=False, **kwargs):
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer/optimizer.py]: can we set the default value of bias_correction to True?

File: src/operator/optimizer_op-inl.h (hunk 10 above)
Comment (access2rohit) [reply src/operator/optimizer_op-inl.h]: @samskalicky These aren't data copies.  FlatTo2D converts the input data to 2D and returns a pointer(https://github.com/apache/incubator-mxnet/blob/master/include/mxnet/tensor_blob.h#L210). The gradient, weights, mean and variance generally have higher dimensions than 2. Refer to this file for examples regarding how other optimizers are tested(https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_optimizer.py)

File: src/operator/optimizer_op-inl.h (hunk 11 above)
Comment (access2rohit) [reply src/operator/optimizer_op-inl.h]: same as above

File: python/mxnet/optimizer/optimizer.py (hunk 12 above)
Comment (access2rohit) [reply python/mxnet/optimizer/optimizer.py]: Earlier we discussed offline to keep it False. Any reason for changing it to True now ?

**Comment by access2rohit** [comment]
> @access2rohit could you answer Sam's comments?

Done

File: src/operator/optimizer_op-inl.h [hunk 13]
```diff
@@ -1563,6 +1563,192 @@ inline void AdamUpdateEx(const nnvm::NodeAttrs& attrs,
...
+      DMLC_DECLARE_FIELD(beta2)
+      .set_default(0.999f)
+      .describe("The decay rate for the 2nd moment estimates.");
+      DMLC_DECLARE_FIELD(epsilon)
+      .set_default(1e-6f)
+      .describe("A small constant for numerical stability.");
+      DMLC_DECLARE_FIELD(t)
+      .describe("Index update count.");
+      DMLC_DECLARE_FIELD(bias_correction)
+      .set_default(false)
```
Comment (eric-haibin-lin) [inline src/operator/optimizer_op-inl.h]: also could you set the default here to True?

File: src/operator/optimizer_op-inl.h [hunk 14]
```diff
@@ -1563,6 +1563,192 @@ inline void AdamUpdateEx(const nnvm::NodeAttrs& attrs,
   }
 }
 
+struct LambUpdatePhaseOneParam : public dmlc::Parameter<LambUpdatePhaseOneParam> {
+    float beta1;
+    float beta2;
+    float epsilon;
+    float t;
```
Comment (sxjscience) [inline src/operator/optimizer_op-inl.h]: @eric-haibin-lin @access2rohit I find this issue when reading the code. Here, the `t` should be the number of updates and should not be stored as `float`, which will lose the precision. I think we need to store it as index_t.

File: src/operator/optimizer_op-inl.h [hunk 15]
```diff
@@ -1563,6 +1563,192 @@ inline void AdamUpdateEx(const nnvm::NodeAttrs& attrs,
...
+    }
+
+    mean_data[i] = beta1 * mean_data[i] + (1.f - beta1) * grad_rescaled;
+    var_data[i] = beta2 * var_data[i] + (1.f - beta2) * grad_rescaled * grad_rescaled;
+
+    DType g = mean_data[i] / (square_root::Map(var_data[i]) + epsilon) + wd * weight_data[i];
+
+    if (bias_correction) {
+      DType mean_hat = mean_data[i] / (1. - power::Map(beta1, t));
+      DType var_hat = var_data[i] / (1 - power::Map(beta2, t));
```
Comment (sxjscience) [inline src/operator/optimizer_op-inl.h]: Actually, in apex, it uses a float32 to calculate the power and then switch to float16:
https://github.com/NVIDIA/apex/blob/325f5a0bec542701edba1628ad34f3b2ea47c556/csrc/multi_tensor_lamb.cu#L231-L249

File: src/operator/optimizer_op-inl.h (hunk 14 above)
Comment (access2rohit) [reply src/operator/optimizer_op-inl.h]: we are using float here for integer data type. @sxjscience can you explain how we will loses precision for the operation `beta^t` ?

**Pull request merged**